slime/backends/fsdp_utils/checkpoint.py

```diff
@@ -0,0 +1,203 @@
+            state_dict["optim"] = flattened_optim
+
+            # Create FQN mapping
+            fqn_to_index = {f"model.{k}": 0 for k in model_state_dict.keys()}
+            fqn_to_index.update({f"optim.{k}": 0 for k in flattened_optim.keys()})
+
+            storage_writer = HuggingFaceStorageWriter(path=checkpoint_dir, fqn_to_index_mapping=fqn_to_index)
+
+            if dist.get_rank() == 0:
+                print("Saving in safetensors format")
+
+        except ImportError:
+            if dist.get_rank() == 0:
+                print("WARNING: safetensors not available, using standard format")
+            storage_writer = dist_cp.FileSystemWriter(checkpoint_dir)
+    else:
+        storage_writer = dist_cp.FileSystemWriter(checkpoint_dir)
+
+    dist_cp.save(state_dict=state_dict, storage_writer=storage_writer)
+
+    # Save tokenizer and training state on rank 0
+    if dist.get_rank() == 0:
+        tokenizer.save_pretrained(checkpoint_dir)
+        torch.save(
+            {"iteration": iteration, "global_step": global_step}, os.path.join(checkpoint_dir, "training_state.pt")
+        )
+        print(f"Checkpoint saved to {checkpoint_dir}")
+
+    dist.barrier()
+
+
+def load_checkpoint(args, model, optimizer):
+    """Load a checkpoint using FSDP distributed checkpointing (FSDP v2 only).
+
+    Returns:
+        (iteration, global_step) - iteration is -1 if no training state found
+    """
+    checkpoint_path = args.load
+    if not checkpoint_path:
+        return -1, 0
+
+    if dist.get_rank() == 0:
+        print(f"Loading checkpoint from {checkpoint_path}")
+
+    # Detect safetensors format
+    is_safetensors = False
+    if dist.get_rank() == 0 and os.path.exists(checkpoint_path):
+        is_safetensors = any(f.endswith(".safetensors") for f in os.listdir(checkpoint_path))
+        if is_safetensors:
+            print("Detected safetensors format")
+
+    is_safetensors_t = torch.tensor([is_safetensors], dtype=torch.long, device="cpu")
+    dist.broadcast(is_safetensors_t, src=0)
+    is_safetensors = bool(is_safetensors_t[0].item())
+
+    # Get empty state dicts with correct structure
+    model_state_dict, optimizer_state_dict = get_state_dict(
+        model, optimizer, options=StateDictOptions(full_state_dict=False, cpu_offload=False)
+    )
+
+    # Load based on format
+    if is_safetensors:
+        try:
+            from torch.distributed.checkpoint import HuggingFaceStorageReader
+
+            state_dict = {"model": model_state_dict, "optim": {}}
+            storage_reader = HuggingFaceStorageReader(path=checkpoint_path)
+            dist_cp.load(state_dict=state_dict, storage_reader=storage_reader)
+
+            # Unflatten optimizer state
+            optimizer_state_dict = _unflatten_optimizer_state(state_dict["optim"])
+
+        except ImportError:
+            if dist.get_rank() == 0:
+                print("WARNING: safetensors reader not available, using standard reader")
```

Comment:
If the `is_safetensors` is True, that means there is only `safetensor` files in the folder, you should raise an error.